src/utils/reality-question-validation/index.ts

```diff
@@ -0,0 +1,112 @@
+import { Address, concat, Hex, keccak256 } from "viem";
+import { LogNewQuestion } from "../../services/reality";
+import { SnapshotProposal } from "../../services/snapshot";
+import { calculateTxHash } from "./eip-712-transaction-hash";
+import {
+  BaseProposalValidationError,
+  ProposalIdMismatchError,
+  SafeHashCalculationError,
+  SafeHashMismatchError,
+  SafeNotFoundForProposalNetworkError,
+  SafeSnapPluginConfigurationError,
+  TxHashMismatchError,
+  ValidationErrorSeverity,
+} from "./errors";
+
+/**
+ * Validate the provided question against the proposal, throwing an error
+ * if anything is incorrect
+ *
+ * @param event - The event to validate.
+ * @param proposal - The proposal associated with the event.
+ *
+ * @throws {BaseProposalValidationError} If the validation fails.
+ *
+ * @example
+ * // This will throw an error if anything is incorrect, otherwise
+ * // it will just execute without returning anything
+ * assertValidRealityQuestion(event, proposal);
+ */
+export const assertValidRealityQuestion = (event: LogNewQuestion, proposal: SnapshotProposal) => {
+  if (proposal.id != event.question.proposalId) {
+    throw new ProposalIdMismatchError(proposal.id, event.question.proposalId);
+  }
+  const safes = proposal.plugins?.safeSnap?.safes;
+  if (!safes || safes.length < 1) {
+    throw new SafeSnapPluginConfigurationError();
+  }
+
+  safes.forEach((safe) => {
+    const { network, realityAddress, txs, hash } = safe;
+
+    const domain = {
+      chainId: Number(network),
+      verifyingContract: realityAddress as Address,
+    } as const;
+
+    const txHashes = txs.map((tx) => {
+      const {
+        hash,
+        mainTransaction: { to, data, nonce, value, operation },
+      } = tx;
+
+      const calculatedHash = calculateTxHash(domain, {
+        to: to as Address,
+        data: data as Hex,
+        nonce: BigInt(nonce),
+        value: BigInt(value),
+        operation: Number(operation),
+      });
+
+      if (hash != calculatedHash) throw new TxHashMismatchError(hash, calculatedHash);
```

Comment:
This kind of error shouldn't happen, but as we are using Snapshot GraphQL instead of an on-chain contract, I validated the integrity of the result.